Generic pointer

**V version:**  V 0.2.4 14073ac.87dfb3e
**OS:** windows, Microsoft Windows 10 Pro v19042 64-bit

**What did you do?**
```v
module main

fn foo<T>(val T) ?T {
	return val
}

struct Bar {
	num int
}

fn main() {
	foo<&Bar>(&Bar{num: 123}) or { panic(err) }
}
```

**What did you expect to see?**
result or readable error
 
**What did you see instead?**
```
==================
C:/Users/russo/AppData/Local/Temp/v_0/v_map.18435184857299524434.tmp.c:444: warning: WINVER redefined
C:/Users/russo/AppData/Local/Temp/v_0/v_map.18435184857299524434.tmp.c:1905: error: ';' expected (got "*")
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.
```

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -3852,7 +3852,7 @@ fn (mut g Gen) return_stmt(node ast.Return) {
 		return
 	}
 	tmpvar := g.new_tmp_var()
-	ret_typ := g.typ(g.fn_decl.return_type)
+	ret_typ := g.typ(g.unwrap_generic(g.fn_decl.return_type))
 	mut use_tmp_var := g.defer_stmts.len > 0 || g.defer_profile_code.len > 0
 	// handle promoting none/error/function returning 'Option'
 	if fn_return_is_optional {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -214,7 +214,7 @@ fn (mut g Gen) gen_fn_decl(node &ast.FnDecl, skip bool) {
 	}
 	//
 	mut name := g.c_fn_name(node) or { return }
-	mut type_name := g.typ(node.return_type)
+	mut type_name := g.typ(g.unwrap_generic(node.return_type))
 
 	if g.pref.obfuscate && g.cur_mod.name == 'main' && name.starts_with('main__') && !node.is_main
 		&& node.name != 'str' {

--- a/vlib/v/tests/generics_with_reference_generic_args_test.v
+++ b/vlib/v/tests/generics_with_reference_generic_args_test.v
@@ -1,0 +1,15 @@
+module main
+
+fn foo<T>(val T) ?T {
+	return val
+}
+
+struct Bar {
+	num int
+}
+
+fn test_generics_with_reference_generic_args() {
+	ret := foo<&Bar>(&Bar{ num: 123 }) or { panic(err) }
+	println(ret)
+	assert ret.num == 123
+}

